src/core/ix86/iR3000A.cc

```diff
@@ -2366,53 +2335,70 @@ void X86DynaRecCPU::recSRAV() {
-        gen.MOV32RtoM((uint32_t)&m_psxRegs.GPR.r[_Rd_], PCSX::ix86::EAX);
+        gen.sar(eax, cl);
+        gen.mov(dword [&m_psxRegs.GPR.r[_Rd_]], eax);
     } else {
         m_iRegs[_Rd_].state = ST_UNK;
 
-        gen.MOV32MtoR(PCSX::ix86::EAX, (uint32_t)&m_psxRegs.GPR.r[_Rt_]);
-        gen.MOV32MtoR(PCSX::ix86::ECX, (uint32_t)&m_psxRegs.GPR.r[_Rs_]);
-        // gen.AND32ItoR(PCSX::ix86::ECX,0x1f);  // MIPS spec says that the shift amount is masked by 31. however this
+        gen.mov(eax, dword [&m_psxRegs.GPR.r[_Rt_]]);
+        gen.mov(ecx, dword [&m_psxRegs.GPR.r[_Rs_]]);
+        // gen.and_(ecx,0x1f);  // MIPS spec says that the shift amount is masked by 31. however this
         // happens implicitly on all x86 processors except for 8086.
         // So no need to do it manually
-        gen.SAR32CLtoR(PCSX::ix86::EAX);
-        gen.MOV32RtoM((uint32_t)&m_psxRegs.GPR.r[_Rd_], PCSX::ix86::EAX);
+        gen.sar(eax, cl);
+        gen.mov(dword [&m_psxRegs.GPR.r[_Rd_]], eax);
     }
 }
 
-void X86DynaRecCPU::recSYSCALL() {
-    gen.MOV32ItoM((uint32_t)&m_psxRegs.pc, (uint32_t)m_pc - 4);
-    gen.MOV32ItoR(PCSX::ix86::EBP, 0xffffffff);
-    gen.MOV32ItoM((uint32_t)&m_arg2, m_inDelaySlot ? 1 : 0);
-    gen.MOV32ItoM((uint32_t)&m_arg1, 0x20);
-    gen.MOV32ItoM((uint32_t)&m_functionPtr, 0);
+/// eax: scratch
+/// ecx: Status Register
+/// ebp: PC after the exception
+void X86DynaRecCPU::recException(uint32_t cause) {
+    // If in delay slot: epc = address of branch, else epc = address of last instruction
+    const uint32_t epc = m_inDelaySlot ? m_pc - 8 : m_pc - 4;
+    if (m_inDelaySlot) {
+        cause |= 0x80000000; // Set BD bit
+    }
 
-    m_pcInEBP = true;
-    m_stopRecompile = true;
+    gen.mov(dword [&m_psxRegs.CP0.n.EPC], epc); // set EPC
+    gen.mov(dword [&m_psxRegs.CP0.n.Cause], cause); // set CAUSE
+
+    gen.mov(ebp, 0x80000000); // ebp = pc after exception
+    gen.mov(eax, 0xBFC00180); // PC if the BEV bit is set in SR
+    gen.mov(ecx, dword [&m_psxRegs.CP0.n.Status]); // ecx = SR
+    gen.test(ecx, 1 << 22); // Check if bev is enabled
+    gen.cmovnz(ebp, eax); // If it is, copy eax to ebp
+    gen.mov(eax, ecx); // Copy SR to eax
+
+    // ASM equivalent to Status = (Status & ~0x3F) | ((Status & 0xF) << 2);
+    gen.and_(eax, 0xF);
+    gen.shl(eax, 2);
+    gen.and_(ecx, ~0x3F);
+    gen.or_(ecx, eax);
+    gen.mov(dword [&m_psxRegs.CP0.n.Status], ecx); // Finally, store SR back
+
+    m_pcInEBP = true; // The PC after the exception is now in EBP
+    m_stopRecompile = true; // Stop compilation (without a delay slot, as exceptions have none)
 }
 
-void X86DynaRecCPU::recBREAK() {
-    gen.MOV32ItoM((uint32_t)&m_psxRegs.pc, (uint32_t)m_pc - 4);
-    gen.MOV32ItoR(PCSX::ix86::EBP, 0xffffffff);
-    gen.MOV32ItoM((uint32_t)&m_arg2, m_inDelaySlot ? 1 : 0);
-    gen.MOV32ItoM((uint32_t)&m_arg1, 0x24);
-    gen.MOV32ItoM((uint32_t)&m_functionPtr, 0);
+void X86DynaRecCPU::recSYSCALL() {
+    recException(0x20);
```

Comment:
Maybe use the exception enum instead? https://github.com/grumpycoders/pcsx-redux/blob/main/src/core/r3000a.h#L295-L306

Reply:
Fixed, thanks